File: Sprint-1/JavaScript/findCommonItems/findCommonItems.js
/**
 * Finds common items between two arrays.
 *
* Time Complexity: O(n × m) — filter visits each of n items in firstArray, and
 *   includes() scans up to m items in secondArray for each one
 * Space Complexity: O(n) — the Set holds at most n items (size of firstArray)
 * Optimal Time Complexity: O(n + m) — convert secondArray to a Set (O(m)), then
 *   each lookup is O(1) instead of O(m), reducing the total to one pass each
 *
 * Refactor: build a Set from secondArray once (O(m)), then filter firstArray
 * using Set.has() (O(1) per lookup) instead of Array.includes() (O(m) per lookup).
 * This changes the nested O(n × m) work into linear O(n + m).
 *
 * @param {Array} firstArray - First array to compare
 * @param {Array} secondArray - Second array to compare
 * @returns {Array} Array containing unique common items
 */
export const findCommonItems = (firstArray, secondArray) => {
  const secondSet = new Set(secondArray);
  return [...new Set(firstArray.filter((item) => secondSet.has(item)))];
};

File: Sprint-1/Python/find_common_items/find_common_items.py
from typing import List, Sequence, TypeVar

ItemType = TypeVar("ItemType")


def find_common_items(
    first_sequence: Sequence[ItemType], second_sequence: Sequence[ItemType]
) -> List[ItemType]:
    """
    Find common items between two arrays.

    Time Complexity: O(n × m) — outer loop visits n items in first_sequence;
        inner loop scans up to m items in second_sequence for each one;
        the `i not in common_items` check adds another O(k) per match
    Space Complexity: O(n) — common_items holds at most n elements
    Optimal time complexity: O(n + m) — convert second_sequence to a Set (O(m)),
        then each membership check is O(1) instead of O(m)

    Refactor: build a set from second_sequence once (O(m)), then iterate
    first_sequence once using O(1) set lookup. Uses a result set to deduplicate.
    Trades a small amount of extra space for linear time overall.
    """

    second_set = set(second_sequence)
    seen: set = set()

    common_items: List[ItemType] = []
    for item in first_sequence:
       if item in second_set and item not in seen:
          seen.add(item)
          common_items.append(item)
    return common_items
